### PR TITLE
splice.nix: add convenience functions 

### PR DESCRIPTION
**Pull request by Artturin**
Replaces https://github.com/NixOS/nixpkgs/pull/105374

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/splice.nix [hunk 1]
```diff
@@ -120,12 +140,28 @@ in
...
-  callPackage = pkgs.newScope {};
+  callPackage = pkgs.newScope { };
 
   callPackages = lib.callPackagesWith splicedPackagesWithXorg;
 
   newScope = extra: lib.callPackageWith (splicedPackagesWithXorg // extra);
 
+  makeScopeWithSplicing = lib.makeScopeWithSplicing splicePackages pkgs.newScope;
+
+  topProjectedSpliced = attr:
```
Comment (Artturin) [inline pkgs/top-level/splice.nix:151]: ideas for a better name?

Comment (Madouura) [reply pkgs/top-level/splice.nix:151]: ``topSplices``, ``topProjectedSplices`` or ``projectedSplices``?
I'm not sure ``top`` really needs to be there.

File: pkgs/development/interpreters/python/default.nix [hunk 2]
```diff
@@ -71,9 +70,7 @@
             overrides
           ]);
           aliases = self: super: lib.optionalAttrs config.allowAliases (import ../../../top-level/python-aliases.nix lib self super);
-        in lib.makeScopeWithSplicing
-          splicePackages
-          newScope
+        in makeScopeWithSplicing
           otherSplices
```
Comment (FRidh) [inline pkgs/development/interpreters/python/default.nix:74]: I think `topProjectedSpliced` can be used here as well.

**Review by FRidh — COMMENTED** [review]
Maybe use a single parameter name for all these `steamPackagesAttr`, `pythonAttr`, ... such as `selfAttr`?

Also, some spliced packages don't expose this attribute. That means it is not possible to *add a variant* through overriding as the overridden version will use the attribute corresponding to the original variant.

**Comment by symphorien** [comment]
could you document what you add? And if you document makeScopeWithSplicing at the same time it would be wonderful.

File: pkgs/top-level/splice.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/top-level/splice.nix]: renamed to generateSplicesForMkScope

File: pkgs/development/interpreters/python/default.nix (hunk 2 above)
Comment (Artturin) [reply pkgs/development/interpreters/python/default.nix:74]: it's fine the way it is with `pythonOnXForX)

**Comment by Ericson2314** [comment]
I just lost some time because this broke on a dot in an attribute name. Yes, that is a preverse edge case, but `splitAttrByPath` taking a list is a *good* thing, and I think we should just do that rather than try to pas a string with `dot` notation.

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/pull/298849/commits/e7ea78e50dc5ec3eae324f26603946344ad9258a this commit (part of an not yet merged PR) make it take a list too

**Pull request merged**